src/Microsoft.AspNetCore.SignalR.Client/HubConnectionBuilderHttpExtensions.cs

```diff
@@ -2,283 +2,42 @@
-        public static readonly string ClientCertificatesKey = "ClientCertificates";
-        public static readonly string CredentialsKey = "Credentials";
-        public static readonly string UseDefaultCredentialsKey = "UseDefaultCredentials";
-
-        public static IHubConnectionBuilder WithUrl(this IHubConnectionBuilder hubConnectionBuilder, string url)
+        public static IHubConnectionBuilder WithHttpConnection(this IHubConnectionBuilder hubConnectionBuilder, Action<HttpConnectionOptions> configureHttpConnection)
         {
-            if (url == null)
-            {
-                throw new ArgumentNullException(nameof(url));
-            }
-
-            return hubConnectionBuilder.WithUrl(new Uri(url));
-        }
+            HttpConnectionOptions options = new HttpConnectionOptions();
+            configureHttpConnection(options);
 
-        public static IHubConnectionBuilder WithUrl(this IHubConnectionBuilder hubConnectionBuilder, Uri url)
-        {
-            if (url == null)
+            var httpOptions = new HttpOptions
             {
-                throw new ArgumentNullException(nameof(url));
-            }
+                HttpMessageHandlerFactory = options.MessageHandlerFactory,
+                Headers = options._headers != null ? new ReadOnlyDictionary<string, string>(options._headers) : null,
+                AccessTokenFactory = options.AccessTokenFactory,
+                WebSocketOptions = options.WebSocketOptions,
+                Cookies = options._cookies,
+                Proxy = options.Proxy,
+                UseDefaultCredentials = options.UseDefaultCredentials,
+                ClientCertificates = options._clientCertificates,
+                Credentials = options.Credentials,
+            };
 
-            hubConnectionBuilder.ConfigureConnectionFactory(() =>
-            {
-                var headers = hubConnectionBuilder.GetHeaders();
-                var httpOptions = new HttpOptions
-                {
-                    HttpMessageHandler = hubConnectionBuilder.GetMessageHandler(),
-                    Headers = headers != null ? new ReadOnlyDictionary<string, string>(headers) : null,
-                    AccessTokenFactory = hubConnectionBuilder.GetAccessTokenFactory(),
-                    WebSocketOptions = hubConnectionBuilder.GetWebSocketOptions(),
-                    Cookies = hubConnectionBuilder.GetCookies(),
-                    Proxy = hubConnectionBuilder.GetProxy(),
-                    UseDefaultCredentials = hubConnectionBuilder.GetUseDefaultCredentials(),
-                    ClientCertificates = hubConnectionBuilder.GetClientCertificates(),
-                    Credentials = hubConnectionBuilder.GetCredentials(),
-                };
+            Func<IConnection> createConnection = () => new HttpConnection(options.Url,
+                options.Transport ?? TransportType.All,
+                null, // TODO: Pass in logger factory
```

Comment:
Thoughts on what to do here?

Reply:
> Get the logger descriptor from the builder's ServiceCollection?

Definately not. We should defer it all. 

```
services.AddSingleton<Func<IConnection>>(serviceProvider =>
{
     return () => new HttpConnection(....)
});
```

Reply:
If we had `IConnectionFactory` as an interface, we could just accept everything as constructor parameters to `HttpConnectionFactory`. Then we wouldn't need a delegate, we'd just register `HttpConnectionFactory` with `TryAdd` and `IOptions<HttpConnectionOptions>` with `Configure`.

Reply:
```
services.AddSingleton<Func<IConnection>>(serviceProvider =>
{
     return () => new HttpConnection(....)
});
```

Yeah I think that would work. The only thing is a func returning a func does feel a little inside baseball for anyone who wants to hook in their own IConnection.

An IConnectionFactory might simply things, but it would need access to the `ServiceCollection` when creating the IConnection instance to get the logger factory. How would that work? Set a logger on `HttpConnectionOptions` and get it that way?

Reply:
> but it would need access to the ServiceCollection when creating the IConnection instance to get the logger factory

Why? It could accept `ILoggerFactory` in it's constructor and stash it away to use when constructing the connection. If necessary, it could also accept `IServerProvider` in it's constructor to get the container itself.